### ISSUE TITLE
get_state_dict cannot save checkpoints of FSDP with local shards 

### ISSUE THREAD

**Issue by fingertap** [pytorch/pytorch#141799]
### 🐛 Describe the bug

When saving checkpoints of FSDP models with local shards, the `get_state_dict` throws not implemented error.

The code below reproduces the error on my machine:

```python
import torch
import torch.nn as nn
import torch.optim as optim
import torch.distributed as dist

import functools
from torch.distributed.fsdp import (
    FullyShardedDataParallel as FSDP,
    StateDictType,
    ShardingStrategy,
)
from torch.distributed.checkpoint.state_dict import get_state_dict
from torch.distributed.fsdp.wrap import size_based_auto_wrap_policy


class Model(nn.Module):
    def __init__(self, size: int):
        super().__init__()
        self.layer1 = nn.Linear(size // 4, 1, bias=False)
        self.layer2 = nn.Linear(1, size // 4, bias=False)
        self.layer3 = nn.Linear(size // 4, 1, bias=False)
        self.layer4 = nn.Linear(1, size // 4, bias=False)

    def forward(self, x):
        x = self.layer1(x)
        x = self.layer2(x)
        x = self.layer3(x)
        x = self.layer4(x)
        return x


def get_auto_wrap_policy():
    return functools.partial(
        size_based_auto_wrap_policy,
        min_num_params=100_000_000,  # 自动包装大于100M参数的层
        force_leaf_modules=set(),
        exclude_wrap_modules=set()
    )


if __name__ == "__main__":
    dist.init_process_group(backend="nccl")
    rank = dist.get_rank()
    world_size = dist.get_world_size()
    torch.cuda.set_device(rank)

    num_devices = 8
    shard_size = 4

    shard_groups = tuple(
        dist.new_group(backend="nccl", ranks=list(range(group, group + shard_size)))
        for group in range(0, num_devices, shard_size)
    )
    current_shard_group = shard_groups[rank // shard_size]

    replicate_groups = tuple(
        dist.new_group(backend="nccl", ranks=list(range(sz, world_size, shard_size)))
        for sz in range(shard_size)
    )
    current_replicate_group = replicate_groups[rank % shard_size]

    size_in_GB = 4
    model_size = size_in_GB * 1024 * 1024 * 1024 // 4
    if rank % shard_size == 0:
        with torch.device("cuda"):
            model = Model(model_size)
        model_init_fn = None
    else:
        with torch.device("meta"):
            model = Model(model_size)
        model_init_fn = lambda x: x.to_empty(
            device=torch.cuda.current_device(), recurse=False
        )


    model = FSDP(
        model,
        process_group=(
            current_shard_group,
            current_replicate_group
        ),
        sharding_strategy=ShardingStrategy.HYBRID_SHARD,
        param_init_fn=model_init_fn,
        auto_wrap_policy=get_auto_wrap_policy(),
        sync_module_states=True,
    )

    # This is ok, yet it gives deprecated warning
    # with FSDP.state_dict_type(model, StateDictType.FULL_STATE_DICT):
    #     state_dict = model.state_dict()

    optimizer = optim.Adam(model.parameters(), lr=0.001)
    # This will raise error
    state_dict, optimizer = get_state_dict(model, optimizer)

    if rank == 0:
        from safetensors.torch import save_file
        save_file(state_dict, f"./model.safetensors")

    for pg in shard_groups:
        dist.destroy_process_group(pg)
    for pg in replicate_groups:
        dist.destroy_process_group(pg)
    dist.destroy_process_group()
```

The error message:

```shell
[rank5]: Traceback (most recent call last):
[rank5]:   File "/workspace/inference/scripts/test_implement/tests/test_fsdp_save_checkpoint.py", line 94, in <module>
[rank5]:     state_dict, optimizer = get_state_dict(model, optimizer)
[rank5]:   File "/home/user/miniconda3/envs/inferenc/lib/python3.10/site-packages/torch/distributed/checkpoint/state_dict.py", line 1109, in get_state_dict
[rank5]:     model_state_dict = _get_model_state_dict(model, info)
[rank5]:   File "/home/user/miniconda3/envs/inference/lib/python3.10/site-packages/torch/distributed/checkpoint/state_dict.py", line 520, in _get_model_state_dict
[rank5]:     if torch.is_tensor(p) and p.is_meta:
[rank5]:   File "/home/user/miniconda3/envs/inference/lib/python3.10/site-packages/typing_extensions.py", line 2853, in wrapper
[rank5]:     return arg(*args, **kwargs)
[rank5]:   File "/home/user/miniconda3/envs/inference/lib/python3.10/site-packages/torch/distributed/_shard/sharded_tensor/api.py", line 1140, in __torch_function__
[rank5]:     return dispatch(st_instance, func)
[rank5]:   File "/home/user/miniconda3/envs/inference/lib/python3.10/site-packages/torch/distributed/_shard/sharded_tensor/api.py", line 1122, in dispatch
[rank5]:     return _SHARDED_OPS[func](types, args, kwargs, st._process_group)
[rank5]:   File "/home/user/miniconda3/envs/inference/lib/python3.10/site-packages/torch/distributed/_shard/op_registry_utils.py", line 33, in wrapper
[rank5]:     return wrapped_func(types, args, kwargs, process_group)
[rank5]:   File "/home/user/miniconda3/envs/inference/lib/python3.10/site-packages/torch/distributed/_shard/sharded_tensor/_ops/tensor_ops.py", line 57, in st_is_meta
[rank5]:     return args[0].local_tensor().is_meta
[rank5]:   File "/home/user/miniconda3/envs/inference/lib/python3.10/site-packages/torch/distributed/_shard/sharded_tensor/api.py", line 1099, in local_tensor
[rank5]:     raise NotImplementedError("Only single local shard is supported.")
[rank5]: NotImplementedError: Only single local shard is supported.
```

Using the conventional way of saving checkpoints throws deprecated warning.

Also, the current interface of `get_state_dict` forces me to pass an optimizer, which I think should support to be `None`.

### Versions

PyTorch version: 2.4.0+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04 LTS (x86_64)
GCC version: (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0
Clang version: Could not collect
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.10.15 (main, Oct  3 2024, 07:27:34) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.4.0-153-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.2.140
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA A800-SXM4-80GB
GPU 1: NVIDIA A800-SXM4-80GB
GPU 2: NVIDIA A800-SXM4-80GB
GPU 3: NVIDIA A800-SXM4-80GB
GPU 4: NVIDIA A800-SXM4-80GB
GPU 5: NVIDIA A800-SXM4-80GB
GPU 6: NVIDIA A800-SXM4-80GB
GPU 7: NVIDIA A800-SXM4-80GB

Nvidia driver version: 535.161.08
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   46 bits physical, 57 bits virtual
Byte Order:                      Little Endian
CPU(s):                          128
On-line CPU(s) list:             0-95
Off-line CPU(s) list:            96-127
Vendor ID:                       GenuineIntel
Model name:                      Intel(R) Xeon(R) Platinum 8358 CPU @ 2.60GHz
CPU family:                      6
Model:                           106
Thread(s) per core:              2
Core(s) per socket:              32
Socket(s):                       2
Stepping:                        6
Frequency boost:                 enabled
CPU max MHz:                     3400.0000
CPU min MHz:                     800.0000
BogoMIPS:                        5200.00
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 invpcid_single ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local wbnoinvd dtherm ida arat pln pts avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq rdpid md_clear pconfig flush_l1d arch_capabilities
Virtualization:                  VT-x
L1d cache:                       3 MiB (64 instances)
L1i cache:                       2 MiB (64 instances)
L2 cache:                        80 MiB (64 instances)
L3 cache:                        96 MiB (2 instances)
NUMA node(s):                    2
NUMA node0 CPU(s):               0-31,64-95
NUMA node1 CPU(s):               32-63,96-127
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Vulnerable: Clear CPU buffers attempted, no microcode; SMT vulnerable
Vulnerability Retbleed:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Retpolines, IBPB conditional, IBRS_FW, STIBP conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected

Versions of relevant libraries:
[pip3] flake8==7.1.1
[pip3] flashinfer==0.1.6+cu121torch2.4
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.26.4
[pip3] nvidia-cublas-cu12==12.1.3.1
[pip3] nvidia-cuda-cupti-cu12==12.1.105
[pip3] nvidia-cuda-nvrtc-cu12==12.1.105
[pip3] nvidia-cuda-runtime-cu12==12.1.105
[pip3] nvidia-cudnn-cu12==9.1.0.70
[pip3] nvidia-cufft-cu12==11.0.2.54
[pip3] nvidia-curand-cu12==10.3.2.106
[pip3] nvidia-cusolver-cu12==11.4.5.107
[pip3] nvidia-cusparse-cu12==12.1.0.106
[pip3] nvidia-nccl-cu12==2.20.5
[pip3] nvidia-nvjitlink-cu12==12.6.85
[pip3] nvidia-nvtx-cu12==12.1.105
[pip3] torch==2.4.0
[pip3] torchao==0.6.1
[pip3] torchmetrics==1.5.1
[pip3] torchvision==0.19.0
[pip3] triton==3.0.0
[conda] flashinfer                0.1.6+cu121torch2.4          pypi_0    pypi
[conda] nvidia-cublas-cu12        12.1.3.1                 pypi_0    pypi
[conda] nvidia-cuda-cupti-cu12    12.1.105                 pypi_0    pypi
[conda] nvidia-cuda-nvrtc-cu12    12.1.105                 pypi_0    pypi
[conda] nvidia-cuda-runtime-cu12  12.1.105                 pypi_0    pypi
[conda] nvidia-cudnn-cu12         9.1.0.70                 pypi_0    pypi
[conda] nvidia-cufft-cu12         11.0.2.54                pypi_0    pypi
[conda] nvidia-curand-cu12        10.3.2.106               pypi_0    pypi
[conda] nvidia-cusolver-cu12      11.4.5.107               pypi_0    pypi
[conda] nvidia-cusparse-cu12      12.1.0.106               pypi_0    pypi
[conda] nvidia-nccl-cu12          2.20.5                   pypi_0    pypi
[conda] nvidia-nvjitlink-cu12     12.6.85                  pypi_0    pypi
[conda] nvidia-nvtx-cu12          12.1.105                 pypi_0    pypi
[conda] torch                     2.4.0                    pypi_0    pypi
[conda] torchao                   0.6.1                    pypi_0    pypi
[conda] triton                    3.0.0                    pypi_0    pypi

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @LucasLLC @pradeepfn

**Comment by fegin** [comment]
@fingertap You can use `get_model_state_dict()` to avoid passing `optim`.

**Comment by fegin** [comment]
I noticed that you use `FULL_STATE_DICT` but doesn't use the same option for `get_state_dict`. Can you change `state_dict, optimizer = get_state_dict(model, optimizer)` to `state_dict = get_model_state_dict(model, options=StateDictOptions(full_state_dict=True))`.

**Comment by fingertap** [comment]
> ```python
> from torch.distributed.checkpoint.state_dict import get_state_dict
> ```

This works, yet still producing a lot warnings. 
```text
/data/miniconda3/envs/vllm/lib/python3.10/site-packages/torch/distributed/fsdp/fully_sharded_data_parallel.py:689: FutureWarning: FSDP.state_dict_type() and FSDP.set_state_dict_type() are being deprecated. Please use APIs, get_state_dict() and set_state_dict(), which can support different parallelisms, FSDP1, FSDP2, DDP. API doc: https://pytorch.org/docs/stable/distributed.checkpoint.html#torch.distributed.checkpoint.state_dict.get_state_dict .Tutorial: https://pytorch.org/tutorials/recipes/distributed_checkpoint_recipe.html .
```

**Comment by fegin** [comment]
okay, this is a real problem. `get_model_state_dict` should capture and ignore the warning. I'll take a look at this problem. Thanks for reporting.

**Comment by fegin** [comment]
I could not reproduce the warning issue. Can you verify if it is actually from `get_model_state_dict` or is it from `FSDP.state_dict_type`? You don't need to call `FSDP.state_dict_type` when using `get_model_state_dict`.

**Comment by fingertap** [comment]
This code should be able to reproduce the warnings:
```python
import torch
import torch.nn as nn
import torch.distributed as dist

import functools
from torch.distributed.fsdp import (
    FullyShardedDataParallel as FSDP,
    ShardingStrategy,
)
from torch.distributed.checkpoint.state_dict import get_state_dict
from torch.distributed.fsdp.wrap import size_based_auto_wrap_policy


class Model(nn.Module):
    def __init__(self, size: int):
        super().__init__()
        self.layer1 = nn.Linear(size // 4, 1, bias=False)
        self.layer2 = nn.Linear(1, size // 4, bias=False)
        self.layer3 = nn.Linear(size // 4, 1, bias=False)
        self.layer4 = nn.Linear(1, size // 4, bias=False)

    def forward(self, x):
        x = self.layer1(x)
        x = self.layer2(x)
        x = self.layer3(x)
        x = self.layer4(x)
        return x


def get_auto_wrap_policy():
    return functools.partial(
        size_based_auto_wrap_policy,
        min_num_params=100_000_000,
        force_leaf_modules=set(),
        exclude_wrap_modules=set(),
    )


if __name__ == "__main__":
    dist.init_process_group(backend="nccl")
    rank = dist.get_rank()
    world_size = dist.get_world_size()
    torch.cuda.set_device(rank)

    num_devices = 8
    shard_size = 4

    shard_groups = tuple(
        dist.new_group(backend="nccl", ranks=list(range(group, group + shard_size)))
        for group in range(0, num_devices, shard_size)
    )
    current_shard_group = shard_groups[rank // shard_size]

    replicate_groups = tuple(
        dist.new_group(backend="nccl", ranks=list(range(sz, world_size, shard_size)))
        for sz in range(shard_size)
    )
    current_replicate_group = replicate_groups[rank % shard_size]

    size_in_GB = 4
    model_size = size_in_GB * 1024 * 1024 * 1024 // 4
    if rank % shard_size == 0:
        with torch.device("cuda"):
            model = Model(model_size)
        model_init_fn = None
    else:
        with torch.device("meta"):
            model = Model(model_size)
        model_init_fn = lambda x: x.to_empty(
            device=torch.cuda.current_device(), recurse=False
        )

    model = FSDP(
        model,
        process_group=(current_shard_group, current_replicate_group),
        sharding_strategy=ShardingStrategy.HYBRID_SHARD,
        param_init_fn=model_init_fn,
        auto_wrap_policy=get_auto_wrap_policy(),
        sync_module_states=True,
    )

    from torch.distributed.checkpoint.state_dict import get_model_state_dict
    from torch.distributed.checkpoint.state_dict import StateDictOptions

    # This gives me a warning
    state_dict = get_model_state_dict(model, options=StateDictOptions(full_state_dict=True))

    for pg in shard_groups:
        dist.destroy_process_group(pg)
    for pg in replicate_groups:
        dist.destroy_process_group(pg)
    dist.destroy_process_group()
```
I ran with 8 A800s: `torchrun --nproc_per_node 8 my_script.py`

**Comment by fingertap** [comment]
I think it's raised here:
<img width="431" alt="image" src="https://github.com/user-attachments/assets/53082648-a2df-4b10-96e2-65e4eb3cf9c8">
I can suppress the warnings in this way. Current context manager "with warnings.catch_warnings" does not work.
I checked that the latest code has fixed this issue:
https://github.com/pytorch/pytorch/blob/9286c21b2283f86655dc5309f8b6b0c916d9518f/torch/distributed/checkpoint/state_dict.py#L361-L363
It just occurs on `2.4.0` which is installed with `vllm==0.6.3.post1`.

**Comment by fingertap** [comment]
Hi @fegin 
I wonder if I can get a generator of model state dict, in case the model is too large to fit in GPU memory. Can I do this with current torch interface?

**Comment by fegin** [comment]
@fingertap Can we do this with CPU offloading? While it is a good idea to make `get_state_dict` a generator, the design of FSDP1 may make this generator idea error-prune (not sure need some thinking).

**Comment by fingertap** [comment]
Can you point out what can cause errors in a generator fashion weight collection? I checked the source code of FSDP, finding that the FSDP will first call the state_dict of the wrapped module. Then it will loop through the collected state dict to collect the full state dict using all gather. I just modify the state dict construction process and yield from it.

**Comment by fingertap** [comment]
CPU offloading can be slow depending on the hardware. I prefer moving the weights around using shared memory.

**Comment by fegin** [comment]
Yes, that seems to a reasonable modification. However, at this point, we are not planning to make it as a generator as this is not very general. Like where should we yield? Your implementation works for you for FSDP only case. But for a general API, which needs to support different parallelisms, we need to decide whether to yield to users. Similarly, how are we going to do the same for the optimizer state_dict. 

Since offloading state_dict to CPU make sense to most users (because later users will write the cpu state_dict to persistent storages), we would stick this approach. Also, you can also consider sharded state_dict, which is the only state_dict option supported for FSDP2 but you can also opt in to sharded state_dict for FSDP1. This should have a similar effect as your implementation.

**Comment by fingertap** [comment]
This is a very useful case when we want to train multiple models using the same device, e.g. in RL settings.

**Comment by fegin** [comment]
Just out of curiosity, what prevent you from using sharded state_dict?

It is not scalable for a general API to support each individual use case with different implementation and interface. so if we are adding some feature, we would need to make it work with most cases and most uses, which require us to not only support FSDP1 but also different parallelism combinations and also optimizer state dict. So this would require some design even if we decide to support the feature, which will not be done in the near term.

**Comment by fingertap** [comment]
I need to load the weights to another model. I want to reduce the peak memory usage. Say I have an inference engine, which uses another tp strategy (in case the model is very large). Then generally, I need to gather weights from FSDP and send it to the inference engine to load. Here I am concerned with two cases:
1. after all gather the weights on cuda:0, gpu will oom.
    -> currently, we have to offload to CPU to prevent this.
2. after all gather the weights on cuda:0, gpu peak memory can be too large for materializing another inference engine.
    -> I have to make sure to free the memory of state dict during loading.
Supporting a generator-style state dict method can address both cases, making no assumption on external models.

**Comment by fegin** [comment]
Not the best solution for your case, but this what you can do: using sharded state_dict and implement the generator based on the sharded state dict and allgather weights to rank0. So it will act like generator-based solution.

**Comment by fingertap** [comment]
This method sounds less invasive. Will give a try.

**Comment by steveepreston** [comment]
Hey @fingertap
@ppwwyyxx in #136907 said that those warnings should fixed by #132621, But I'm still spammed by `FutureWarning: FSDP.state_dict_type() and FSDP.set_state_dict_type() are being deprecated` warnings in torch `2.5.1+cu124`

**Comment by fegin** [comment]
@steveepreston Do you have a minirepro for the issue?
